ChemistryLib/PhreeqcInterface.cpp

```diff
@@ -0,0 +1,312 @@
+                        break;
+                }
+            }
+
+            if (pv_name == "H")
+            {
+                switch (status)
+                {
+                    case Status::SettingAqueousSolutions:
+                    {
+                        // Set pH value by hydrogen concentration.
+                        auto pH =
+                            -std::log10((*process_solution_iter)->get(node_id));
+                        solution.setpH(pH);
+                        break;
+                    }
+                    case Status::UpdatingProcessSolutions:
+                    {
+                        // Update hydrogen concentration by pH value.
+                        auto hydrogen_concentration =
+                            std::pow(10, -solution.getpH());
+                        (*process_solution_iter)
+                            ->set(node_id, hydrogen_concentration);
+                        break;
+                    }
+                }
+            }
+
+            ++process_id;
+        }
+    }
+}
+
+void phreeqcIO::writeInputsToFile()
+{
+    DBUG("Writing phreeqc inputs into file '%s'.", _input_file);
+    std::ofstream outFile(_input_file, std::ofstream::out);
+
+    if (!outFile.good())
+        ERR("Could not open file %s.", _input_file);
+
+    outFile << *this;
+    outFile.close();
+}
+
+std::ofstream& operator<<(std::ofstream& outFile, phreeqcIO const& phreeqcIO)
+{
+    int const nNodes = phreeqcIO._solutions.size();
+    for (int node_id = 0; node_id < nNodes; ++node_id)
+    {
+        auto const& solution = phreeqcIO._solutions[node_id];
+        outFile << "SOLUTION " << node_id + 1 << "\n";
+        outFile << solution << "\n";
+
+        auto const& nodal_equilibrium_phases =
+            phreeqcIO._equilibrium_phases[node_id];
+        if (!nodal_equilibrium_phases.empty())
+        {
+            outFile << "EQUILIBRIUM_PHASES " << node_id + 1 << "\n";
+            outFile << nodal_equilibrium_phases << "\n";
+        }
+
+        auto const& nodal_kinetic_reactant =
+            phreeqcIO._kinetic_reactants[node_id];
+        if (!nodal_kinetic_reactant.empty())
+        {
+            outFile << "KINETICS " << node_id + 1 << "\n";
+            outFile << nodal_kinetic_reactant;
+            outFile << "-steps " << phreeqcIO._dt << "\n" << "\n";
+        }
+
+        if (node_id == 0)
+        {
+            auto const& reaction_rates = phreeqcIO._reaction_rates;
+            if (!reaction_rates.empty())
+            {
+                outFile << "RATES" << "\n";
+                outFile << reaction_rates << "\n";
+            }
+
+            outFile << "SELECTED_OUTPUT" << "\n";
+            outFile << *phreeqcIO._output;
+        }
+
+        outFile << "END" << "\n" << "\n";
+    }
+
+    return outFile;
+}
+
+void phreeqcIO::execute()
+{
+    auto database_loaded = [&](int instance_id) {
+        if (LoadDatabase(instance_id, _database.c_str()) == 0)
```

Comment:
```c++
        return LoadDatabase(instance_id, _database.c_str()) == 0
```
Result of the operator== is a boolean already...

Reply:
Learned. Fixed.